spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableStrategy.scala

```diff
@@ -0,0 +1,256 @@
+import org.apache.spark.sql.delta.zorder.{ZCube, ZCubeInfo}
+
+import org.apache.spark.sql.SparkSession
+
+object OptimizeTableMode extends Enumeration {
+  type OptimizeTableMode = Value
+  val COMPACTION, ZORDER, CLUSTERING = Value
+}
+
+/**
+ * Defines set of utilities used in OptimizeTableCommand. The behavior of these utilities will
+ * change based on the [[OptimizeTableMode]]: COMPACTION, ZORDER and CLUSTERING.
+ */
+trait OptimizeTableStrategy {
+  def sparkSession: SparkSession
+
+  /**
+   * Utility method to get max bin size in bytes to group files into.
+   */
+  def maxBinSize: Long
+
+  /**
+   * Utility method to prepare files in a partition for optimization.
+   *
+   * By default it sorts files on the size for the binpack.
+   *
+   * @return Prepared files for the subsequent optimization.
+   */
+  def prepareFilesPerPartition(inputFiles: Seq[AddFile]): Seq[AddFile] = inputFiles.sortBy(_.size)
+
+  def optimizeTableMode: OptimizeTableMode.Value
+
+  /**
+   * The clustering algorithm to be used by either by ZORDER or Liquid CLUSTERING.
+   * An error is thrown for COMPACTION.
+   */
+  def curve: String
+
+  /**
+   * During the CLUSTERING mode, establish a new [[ZCubeInfo]] upon the creation of a new ZCUBE.
+   */
+  def createZCubeInfoOpt: Option[ZCubeInfo] = None
+
+  /**
+   * Incorporate essential tags for optimized files based on the [[OptimizeTableMode]].
+   */
+  def tagAddFile(file: AddFile, zCubeInfoOpt: Option[ZCubeInfo]): AddFile =
+    file.copy(dataChange = false)
+
+  /**
+   * Utility to update additional metrics after optimization.
+   *
+   * @param optimizeStats The input stats to update on.
+   * @param removedFiles Removed files.
+   * @param bins Bin-packed file groups from [[groupFilesIntoBins]].
+   */
+  def updateOptimizeStats(
+      optimizeStats: OptimizeStats,
+      removedFiles: Seq[RemoveFile],
+      bins: Seq[(Map[String, String], Seq[AddFile])]): Unit
+}
+
+object OptimizeTableStrategy {
+  def apply(
+      mode: OptimizeTableMode.Value,
+      sparkSession: SparkSession,
+      snapshot: Snapshot,
+      optimizeContext: DeltaOptimizeContext,
+      zOrderBy: Seq[String]): OptimizeTableStrategy = mode match {
+    case OptimizeTableMode.CLUSTERING =>
+      ClusteringOptimizeTableStrategy(
+        sparkSession, ClusteringColumnInfo.extractLogicalNames(snapshot))
+    case OptimizeTableMode.ZORDER => ZOrderOptimizeTableStrategy(sparkSession, zOrderBy)
+    case OptimizeTableMode.COMPACTION =>
+      CompactionOptimizeTableStrategy(sparkSession, optimizeContext)
+    case other => throw new UnsupportedOperationException(s"Unsupported mode $other")
+  }
+
+  def getMode(snapshot: Snapshot, zOrderBy: Seq[String]): OptimizeTableMode.Value = {
+    val isClusteredTable = ClusteredTableUtils.isSupported(snapshot.protocol)
+    val isZOrderBy = zOrderBy.nonEmpty
+    if (isClusteredTable) {
+      assert(!isZOrderBy)
+      OptimizeTableMode.CLUSTERING
```

Comment:
I think if clustering columns are empty, it should fall under compaction right?

Reply:
I think we can do this as a followup given this chage has been large.